// ngx-print-element/esm2015/lib/ngx-print-element.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { NgxPrintElementService } from './ngx-print-element.service';
export class NgxPrintElementComponent {
    /**
     * @param {?} prints
     */
    constructor(prints) {
        this.prints = prints;
    }
    /**
     * print
     * @param {?} id
     * @param {?=} config
     * @return {?}
     */
    print(id, config) {
        this.prints.print(id, config);
    }
}
NgxPrintElementComponent.decorators = [
    { type: Component, args: [{
                selector: 'ngx-print-element',
                template: `<ng-content></ng-content>`,
                exportAs: 'element'
            }] }
];
NgxPrintElementComponent.ctorParameters = () => [
    { type: NgxPrintElementService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    NgxPrintElementComponent.prototype.prints;
}
