// ngx-print-element/esm5/lib/ngx-print-element.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { NgxPrintElementService } from './ngx-print-element.service';
var NgxPrintElementComponent = /** @class */ (function () {
    function NgxPrintElementComponent(prints) {
        this.prints = prints;
    }
    /**
     * print
     * @param id
     * @param config
     */
    /**
     * print
     * @param {?} id
     * @param {?=} config
     * @return {?}
     */
    NgxPrintElementComponent.prototype.print = /**
     * print
     * @param {?} id
     * @param {?=} config
     * @return {?}
     */
    function (id, config) {
        this.prints.print(id, config);
    };
    NgxPrintElementComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ngx-print-element',
                    template: "<ng-content></ng-content>",
                    exportAs: 'element'
                }] }
    ];
    NgxPrintElementComponent.ctorParameters = function () { return [
        { type: NgxPrintElementService }
    ]; };
    return NgxPrintElementComponent;
}());
export { NgxPrintElementComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    NgxPrintElementComponent.prototype.prints;
}
