// ngx-slimscroll/esm5/public_api.js
"use strict";
/**
 * @fileoverview added by tsickle
 * Generated from: public_api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
Object.defineProperty(exports, "__esModule", { value: true });
var slimscroll_event_class_1 = require("./src/classes/slimscroll-event.class");
exports.SlimScrollEvent = slimscroll_event_class_1.SlimScrollEvent;
var slimscroll_options_class_1 = require("./src/classes/slimscroll-options.class");
exports.SLIMSCROLL_DEFAULTS = slimscroll_options_class_1.SLIMSCROLL_DEFAULTS;
exports.SlimScrollOptions = slimscroll_options_class_1.SlimScrollOptions;
var slimscroll_directive_1 = require("./src/directives/slimscroll.directive");
exports.easing = slimscroll_directive_1.easing;
exports.SlimScrollDirective = slimscroll_directive_1.SlimScrollDirective;
var ngx_slimscroll_module_1 = require("./src/module/ngx-slimscroll.module");
exports.NgSlimScrollModule = ngx_slimscroll_module_1.NgSlimScrollModule;
