// ngx-slimscroll/esm5/src/classes/slimscroll-event.class.js
"use strict";
/**
 * @fileoverview added by tsickle
 * Generated from: src/classes/slimscroll-event.class.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @record
 */
function ISlimScrollEvent() { }
exports.ISlimScrollEvent = ISlimScrollEvent;
if (false) {
    /** @type {?} */
    ISlimScrollEvent.prototype.type;
    /** @type {?|undefined} */
    ISlimScrollEvent.prototype.y;
    /** @type {?|undefined} */
    ISlimScrollEvent.prototype.percent;
    /** @type {?|undefined} */
    ISlimScrollEvent.prototype.duration;
    /** @type {?|undefined} */
    ISlimScrollEvent.prototype.easing;
}
var SlimScrollEvent = /** @class */ (function () {
    function SlimScrollEvent(obj) {
        this.type = obj.type;
        this.y = obj && obj.y ? obj.y : 0;
        this.percent = obj && obj.percent ? obj.percent : 0;
        this.duration = obj && obj.duration ? obj.duration : 0;
        this.easing = obj && obj.easing ? obj.easing : 'linear';
    }
    return SlimScrollEvent;
}());
exports.SlimScrollEvent = SlimScrollEvent;
if (false) {
    /** @type {?} */
    SlimScrollEvent.prototype.type;
    /** @type {?} */
    SlimScrollEvent.prototype.y;
    /** @type {?} */
    SlimScrollEvent.prototype.percent;
    /** @type {?} */
    SlimScrollEvent.prototype.duration;
    /** @type {?} */
    SlimScrollEvent.prototype.easing;
}
