// ngx-slimscroll/esm5/src/classes/slimscroll-options.class.js
"use strict";
/**
 * @fileoverview added by tsickle
 * Generated from: src/classes/slimscroll-options.class.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
/**
 * @record
 */
function ISlimScrollOptions() { }
exports.ISlimScrollOptions = ISlimScrollOptions;
if (false) {
    /** @type {?|undefined} */
    ISlimScrollOptions.prototype.position;
    /** @type {?|undefined} */
    ISlimScrollOptions.prototype.barBackground;
    /** @type {?|undefined} */
    ISlimScrollOptions.prototype.barOpacity;
    /** @type {?|undefined} */
    ISlimScrollOptions.prototype.barWidth;
    /** @type {?|undefined} */
    ISlimScrollOptions.prototype.barBorderRadius;
    /** @type {?|undefined} */
    ISlimScrollOptions.prototype.barMargin;
    /** @type {?|undefined} */
    ISlimScrollOptions.prototype.gridBackground;
    /** @type {?|undefined} */
    ISlimScrollOptions.prototype.gridOpacity;
    /** @type {?|undefined} */
    ISlimScrollOptions.prototype.gridWidth;
    /** @type {?|undefined} */
    ISlimScrollOptions.prototype.gridBorderRadius;
    /** @type {?|undefined} */
    ISlimScrollOptions.prototype.gridMargin;
    /** @type {?|undefined} */
    ISlimScrollOptions.prototype.alwaysVisible;
    /** @type {?|undefined} */
    ISlimScrollOptions.prototype.visibleTimeout;
}
/** @type {?} */
exports.SLIMSCROLL_DEFAULTS = new core_1.InjectionToken('NGX_SLIMSCROLL_DEFAULTS');
var SlimScrollOptions = /** @class */ (function () {
    function SlimScrollOptions(obj) {
        this.position = obj && obj.position ? obj.position : 'right';
        this.barBackground = obj && obj.barBackground ? obj.barBackground : '#343a40';
        this.barOpacity = obj && obj.barOpacity ? obj.barOpacity : '1';
        this.barWidth = obj && obj.barWidth ? obj.barWidth : '12';
        this.barBorderRadius = obj && obj.barBorderRadius ? obj.barBorderRadius : '5';
        this.barMargin = obj && obj.barMargin ? obj.barMargin : '1px 0';
        this.gridBackground = obj && obj.gridBackground ? obj.gridBackground : '#adb5bd';
        this.gridOpacity = obj && obj.gridOpacity ? obj.gridOpacity : '1';
        this.gridWidth = obj && obj.gridWidth ? obj.gridWidth : '8';
        this.gridBorderRadius = obj && obj.gridBorderRadius ? obj.gridBorderRadius : '10';
        this.gridMargin = obj && obj.gridMargin ? obj.gridMargin : '1px 2px';
        this.alwaysVisible = obj && typeof obj.alwaysVisible !== 'undefined' ? obj.alwaysVisible : true;
        this.visibleTimeout = obj && obj.visibleTimeout ? obj.visibleTimeout : 1000;
    }
    /**
     * @param {?=} obj
     * @return {?}
     */
    SlimScrollOptions.prototype.merge = /**
     * @param {?=} obj
     * @return {?}
     */
    function (obj) {
        /** @type {?} */
        var result = new SlimScrollOptions();
        result.position = obj && obj.position ? obj.position : this.position;
        result.barBackground = obj && obj.barBackground ? obj.barBackground : this.barBackground;
        result.barOpacity = obj && obj.barOpacity ? obj.barOpacity : this.barOpacity;
        result.barWidth = obj && obj.barWidth ? obj.barWidth : this.barWidth;
        result.barBorderRadius = obj && obj.barBorderRadius ? obj.barBorderRadius : this.barBorderRadius;
        result.barMargin = obj && obj.barMargin ? obj.barMargin : this.barMargin;
        result.gridBackground = obj && obj.gridBackground ? obj.gridBackground : this.gridBackground;
        result.gridOpacity = obj && obj.gridOpacity ? obj.gridOpacity : this.gridOpacity;
        result.gridWidth = obj && obj.gridWidth ? obj.gridWidth : this.gridWidth;
        result.gridBorderRadius = obj && obj.gridBorderRadius ? obj.gridBorderRadius : this.gridBorderRadius;
        result.gridMargin = obj && obj.gridMargin ? obj.gridMargin : this.gridMargin;
        result.alwaysVisible = obj && typeof obj.alwaysVisible !== 'undefined' ? obj.alwaysVisible : this.alwaysVisible;
        result.visibleTimeout = obj && obj.visibleTimeout ? obj.visibleTimeout : this.visibleTimeout;
        return result;
    };
    return SlimScrollOptions;
}());
exports.SlimScrollOptions = SlimScrollOptions;
if (false) {
    /** @type {?} */
    SlimScrollOptions.prototype.position;
    /** @type {?} */
    SlimScrollOptions.prototype.barBackground;
    /** @type {?} */
    SlimScrollOptions.prototype.barOpacity;
    /** @type {?} */
    SlimScrollOptions.prototype.barWidth;
    /** @type {?} */
    SlimScrollOptions.prototype.barBorderRadius;
    /** @type {?} */
    SlimScrollOptions.prototype.barMargin;
    /** @type {?} */
    SlimScrollOptions.prototype.gridBackground;
    /** @type {?} */
    SlimScrollOptions.prototype.gridOpacity;
    /** @type {?} */
    SlimScrollOptions.prototype.gridWidth;
    /** @type {?} */
    SlimScrollOptions.prototype.gridBorderRadius;
    /** @type {?} */
    SlimScrollOptions.prototype.gridMargin;
    /** @type {?} */
    SlimScrollOptions.prototype.alwaysVisible;
    /** @type {?} */
    SlimScrollOptions.prototype.visibleTimeout;
}
